pkg/provisioner/ironic/ironic.go

```diff
@@ -1,62 +1,313 @@
-		p.log.Info("setting provisioning id",
-			"provisioningID", p.host.Status.Provisioning.ID)
+	var ironicNode *nodes.Node
+
+	// Try to load the node by UUID
+	if p.status.ID != "" {
+		// Look for the node to see if it exists (maybe Ironic was
+		// restarted)
+		p.log.Info("looking for existing node by ID", "ID", p.status.ID)
+		ironicNode, err = nodes.Get(p.client, p.status.ID).Extract()
+		switch err.(type) {
+		case nil:
+			p.log.Info("found existing node by ID")
+		case gophercloud.ErrDefault404:
+			p.log.Info("did not find existing node by ID")
+		default:
+			return false, errors.Wrap(err,
+				fmt.Sprintf("failed to find node by ID %s", p.status.ID))
+		}
+	}
+
+	// Try to load the node by name
+	if ironicNode == nil {
+		p.log.Info("looking for existing node by name", "name", p.host.Name)
+		ironicNode, err = nodes.Get(p.client, p.host.Name).Extract()
+		switch err.(type) {
+		case nil:
+			p.log.Info("found existing node by name")
+			// Store the ID so other methods can assume it is set and
+			// so we can find the node using that value next time.
+			p.status.ID = ironicNode.UUID
+			dirty = true
+			p.log.Info("setting provisioning id", "ID", p.status.ID)
+		case gophercloud.ErrDefault404:
+			p.log.Info("did not find existing node by name")
+			ironicNode = nil
+		default:
+			return false, errors.Wrap(err,
+				fmt.Sprintf("failed to find node by name %s", p.host.Name))
+		}
+	}
+
+	// If we have not found a node yet, we need to create one
+	if ironicNode == nil {
+		p.log.Info("registering host in ironic")
+
+		driverInfo := p.bmcAccess.DriverInfo(p.bmcCreds)
+		// FIXME(dhellmann): The names of the images are tied
+		// to the version of ironic we are using and are
+		// likely to change.
+		//
+		// FIXME(dhellmann): We need to get our IP on the
+		// provisioning network from somewhere.
+		driverInfo["deploy_kernel"] = "http://172.22.0.1/images/tinyipa-stable-rocky.vmlinuz"
+		driverInfo["deploy_ramdisk"] = "http://172.22.0.1/images/tinyipa-stable-rocky.gz"
+
+		ironicNode, err = nodes.Create(
+			p.client,
+			nodes.CreateOpts{
+				Driver:        "ipmi",
+				BootInterface: "pxe",
+				Name:          p.host.Name,
+				DriverInfo:    driverInfo,
+			}).Extract()
+		if err != nil {
+			return false, errors.Wrap(err, "failed to register host in ironic")
+		}
+
+		// Store the ID so other methods can assume it is set and so
+		// we can find the node again later.
+		p.status.ID = ironicNode.UUID
 		dirty = true
+		p.log.Info("setting provisioning id", "ID", p.status.ID)
+
+		// If we know the MAC, create a port. Otherwise we will have
+		// to do this after we run the introspection step.
+		if p.host.Spec.BootMACAddress != "" {
+			enable := true
+			p.log.Info("creating port for node in ironic", "MAC",
+				p.host.Spec.BootMACAddress)
+			_, err := ports.Create(
+				p.client,
+				ports.CreateOpts{
+					NodeUUID:   ironicNode.UUID,
+					Address:    p.host.Spec.BootMACAddress,
+					PXEEnabled: &enable,
```

Comment:
If one day we want to support spine/leaf, we'll need the PhysicalNetwork field as well.

Reply:
Let's address that when we get there.

Reply:
++